### ISSUE TITLE
&nbsp; displays with Firefox and getFile .content

### ISSUE THREAD

**Issue by zspecza** [OscarGodson/EpicEditor#204]
Playing around with 0.2.1 along with syncing with a `<textarea>` and syntax highlighting via highlight.js, saving content to a MongoDB database, I find that within code tags, any indentation via spaces is outputted as `&nbsp;` for each space except the last, that one outputs fine. 

I've tried breaking things a little by removing the `&nbsp;`'s from `.replace()` in lines 183, 1377 and 1380 of epiceditor.js, but I still see that on output even if I create a new document after the changes... Is really doing my head in. I've also tried looking at highlight.js but there's not even a mention of `&nbsp;` there that I can find with Sublime Text.

This _only_ happens in code tags, nowhere else... Any thoughts?

**Comment by OscarGodson** [comment]
I'm not sure off the top of my head, but I will look into it. I believe highlight.js actually adds those in their code and I'm not sure the sequence of events that happen. However, are you using the textarea option I just added? And does this happen without highlight.js but just manually typing in:

```
<code>
Foo
  Hello
        World
</code>
```

? Also, it'd be awesome if you could upload a testcase or put something on JSBin :)

**Comment by zspecza** [comment]
I am using the textarea option you just added. It happens without highlight.js too. http://jsbin.com/udaboy/1/edit
Seems it works on non-code tags too. That didn't happen on my local instance after converting to html, which was done simply by wrapping `self.getFiles(textareaFileName, true).content` in a `marked()` call on lines 931 and 935.

**Comment by OscarGodson** [comment]
Perfect! Thanks man, I'll take a look at it tonight or tomorrow. This helps a lot.

**Comment by OscarGodson** [comment]
Hmm, ok, having trouble reproing. I put this in the editor. All export functions and the text too doesnt have any `&nbsp;`. Do you have an example of the exact steps or a screencast to repro?

![Screen Shot 2013-01-31 at 5 49 52 PM](https://f.cloud.github.com/assets/20615/116966/dd2ef98e-6c11-11e2-90d3-43ce9060841e.png)

**Comment by zspecza** [comment]
That's strange. Hmmm... 

Have you looked at the JSBin I provided? That reproduces the issue as far as text goes. Would the browser I'm using make any difference? Using the latest Firefox.

**Comment by OscarGodson** [comment]
That could be it. I tried chrome. The screenshot is from your JSBin.

**Comment by zspecza** [comment]
Hardly recognized it (the image) because the basePath I added didn't affect the style of the editor div on my end, which I thought was just a JSBin thing, heh.  I'm downloading Chrome now to see if it's any different to Firefox on my end. If it is then I guess we've nailed what's causing the issue...

**Comment by OscarGodson** [comment]
Yeah, Firefox and chrome handle content editable differently. Hopefully that's it. I'm not near a computer but will try Firefox as well when I can.

**Comment by zspecza** [comment]
It works on Chrome, so this issue only applies to Firefox - but I suspect it can become a quick nuisance to websites that allow every user to add new content using EpicEditor and then storing that data in a DB... everyone using Firefox would become incredibly annoying, lol.

**Comment by OscarGodson** [comment]
Yeah, totally. I'll fix it for sure and add tests :)

**Comment by OscarGodson** [comment]
Just an update, I was able to repro it, added a failing test, but struggling to actually fix it. This will for sure be fixed soon tho and will be in the 0.2.1 release.

**Comment by OscarGodson** [comment]
@tylermolamphy what?! haha

**Comment by zspecza** [comment]
Mhm I've been trying to come up with a fix for this myself - though I'm not a seasoned JS developer... No luck yet though.

**Comment by OscarGodson** [comment]
So, thanks to @lamplightdev your problem will be patched when you pull develop next, but I changed your ticket name. He fixed it by changing the method to get the content. exportFile goes through some cleansing before displaying it whereas getFile does not. However, `getFile` should _should_ give the same raw text as `exportFile`.

**Comment by OscarGodson** [comment]
Note to self: getFiles doesn't have any documentation in the code. Update this as well.

**Comment by Gankra** [comment]
So to clarify, this bug is now just that the text obtained via getFiles is not sanitized? Could you just tweak the structure of the files so that .content is just a getter method that runs the text sanitization?

**Comment by OscarGodson** [comment]
Of the files? The files are stringified so they can't be functions.

**Comment by Gankra** [comment]
Right, so when you call getFiles, after you've loaded them in, move all the .contents to ._content and make .content a getter that sanitizes ._content.

Or always keep content in ._content so you don't need to move stuff around when you load stuff in.

**Comment by OscarGodson** [comment]
Sure, whatever works :) Just make sure whatever method you use that it keeps all the whitespace and also wraps correctly and that on reload the content doesn't shift and works across the browsers the same.

So, make sure:

```
1
  2
    3
  2
1
```

keeps the right whitespace.

Also make sure stuff like

```
Some really long string would go here that wraps to two lines
```

Doesn't get turned into:

```
Some&nbsp;really&nbsp;long&nbsp;string&nbsp;would&nbsp;go&nbsp;here&nbsp;that&nbsp;wraps&nbsp;to&nbsp;two&nbsp;lines
```

Which would prevent the text from wrapping.

**Comment by Gankra** [comment]
I'm having trouble reproducing the issue described, are there precise repro steps and/or a failing unit test written somewhere?

**Comment by OscarGodson** [comment]
No unit tests for this case yet, but in Firefox:

```
foo
  foo
          foo
```

Type something like that. Then:

```
var x = editor.getFiles();
console.log(x);
```

notice that the content contains `&nbsp;`'s

**Comment by Gankra** [comment]
So I've noticed that, at least in my version of the develop branch, editor.exportFile(); seems to strip the \n's from the text in Firefox but not Chrome. Is this correct behaviour?

It means your sample gets crushed to 

`foo  foo          foo` 

in firefox if you just paste it and refresh

**Comment by OscarGodson** [comment]
Not happening for me on Firefox :\ checked against master and develop. I can send a screencast too if you want.

**Comment by Gankra** [comment]
Ah, never mind, it was just an emergent property of #100.

**Comment by OscarGodson** [comment]
Excited to say this is closed and 0.2.1 is ready for QAing and to have the release notes written up. Thanks a lot @Gankro!!!!!
